### DISCUSSION TITLE
new Function and evalUmdBundle

### DISCUSSION THREAD

**Discussion by dagda1** [tiny-frontend/tiny-client#7] [Q&A] [answered]
Hi,

This project looked perfect for my needs but unfortunately, we have a strict content-security-policy and `new Function` will trigger `unsafe-eval`.

Would it be possible to explain what this code is doing or why is the bundle being called like this?

```ts
  try {
    new Function(umdBundleSource)();
  } finally {
    globalThis.define = previousDefine;
  }
  ```

**Comment by adrienbaron** [comment, marked as answer]
Hey @dagda1 !

Thank you for your interest in the project. The main architecture principle behind tiny frontend is to get the content of the bundle at runtime, so that bundles can be deployed independently from the host application. You can read more about the whole architecture [here](https://tiny-frontend.github.io/guide/about.html#before-we-start).

That means at one point, at runtime, your host app will have to fetch some file from the network and evaluate its content dynamically. This is what this `new Function` call does here, there is not really a way around it.
Evals are usually frowned upon for security concerns, but in this case because you're loading the JS from a trusted source (your own backend), and there are no user input involved, it can be considered safe. However in a strict JS environment this sadly doesn't work.

If you only cared about loading the remote bundle in a browser environment, you could also create a `script` tag pointing to the bundle instead of relying on this `new Function` but it's not safer, just a different way of loading the JS, and it won't work on server side (where you don't have a browser to create script tag in).

Hope it all makes sense!

**Comment by dagda1** [comment]
Hi @adrienbaron ,

Thanks for the answer and it makes sense.  Unfortunately I've been through this already with netlify-cms which has a `new Function` and we could not use it at the company I work for as the CSP is non-negotionable and `unsafe-eval` is turned on.

**Reply by adrienbaron** [reply]
Does your CSP allows for dynamic creation of script tags?

I never thought about CSP when making the choice of using `new Function` both on client and server, but taking that into account it might be worth thinking about just using scripts on client and keeping `new Function` for server side 🤔

**Comment by dagda1** [comment]
Creating script tags would be fine if we can add a `nonce` that keeps CSP happy.  I'm  not sure if we would pass one in or if it could be contained in the framework.  Hopefully the latter

**Reply by adrienbaron** [reply]
I have limited knowledge of CSP but from my Googling it seems like `nonce` are used for inline script. In this case it would not be inline but rather a new script tag with an src attribute pointing to the remote bundle that would be created dynamically by the host.
Something like:
```html
<script src="https://yourdomain.com/api/tiny/bundle/myBundle.js" />
```

If `https://yourdomain.com` is a trusted source I'd assume it'd work?

**Comment by dagda1** [comment]
@adrienbaron you are right, I thought you meant inline script

**Reply by adrienbaron** [reply]
Might look into it if I get a moment, I'll post here if I get something done 👍

**Comment by adrienbaron** [comment]
@dagda1 just released [a new version of tiny client](https://github.com/tiny-frontend/tiny-client/releases/tag/0.0.7) that uses script tag insertion instead of eval on client side. This should work with your CSP!
